Reliably handle failing streams during writes

ctiller opened the pull request:
Introduce a robust mechanism for closures that are covering write scheduling in the combiner.

Specifically fixes:
https://grpc-testing.appspot.com/job/gRPC_master_asanitizer_cpp/310/testReport/junit/(root)/c++_linux_asan/bins_asan_end2end_test___gtest_filter_End2end_End2endTest_ClientCancelsBidi_0__GRPC_POLL_STRATEGY_epoll/

ctiller commented:
Merged with head, rerunning tests to see how it looks.

vjpai commented on src/core/ext/transport/chttp2/transport/chttp2_transport.c:
```diff
@@ -868,7 +876,14 @@ void grpc_chttp2_complete_closure_step(grpc_exec_ctx *exec_ctx,
-    grpc_closure_run(exec_ctx, closure, closure->error_data.error);
+    if (t->write_state == GRPC_CHTTP2_WRITE_STATE_IDLE ||
+        (closure->next_data.scratch & CLOSURE_BARRIER_CANNOT_RUN_WITH_WRITE) ==
+            0) {
```
Can I suggest extra parens around each operand of the || operator to avoid reader confusion when mixing with bitwise ops? Also, for a bitmask isn't `!` a more common paradigm than `== 0` ?

ctiller commented:
Updated (and renamed CLOSURE_BARRIER_CANNOT_RUN_WITH_WRITE to CLOSURE_BARRIER_MAY_COVER_WRITE to avoid a not-not)

ASAN test result is here: https://grpc-testing.appspot.com/view/Pull%20requests/job/gRPC_pull_requests_asan_cpp/577/#showFailuresLink
